lib/ts/recipe/webauthn/api/implementation.ts

```diff
@@ -0,0 +1,1095 @@
+                timeout: response.timeout,
+                userVerification: response.userVerification,
+            };
+        },
+
+        signUpPOST: async function ({
+            webauthnGeneratedOptionsId,
+            credential,
+            tenantId,
+            session,
+            shouldTryLinkingWithSessionUser,
+            options,
+            userContext,
+        }: {
+            webauthnGeneratedOptionsId: string;
+            credential: CredentialPayload;
+            tenantId: string;
+            session: SessionContainerInterface | undefined;
+            shouldTryLinkingWithSessionUser: boolean | undefined;
+            options: APIOptions;
+            userContext: UserContext;
+            // should also have the email or recoverAccountToken in order to do the preauth checks
+        }): Promise<
+            | {
+                  status: "OK";
+                  session: SessionContainerInterface;
+                  user: User;
+              }
+            | GeneralErrorResponse
+            | {
+                  status: "SIGN_UP_NOT_ALLOWED";
+                  reason: string;
+              }
+            | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+            | { status: "WRONG_CREDENTIALS_ERROR" }
+            | { status: "INVALID_AUTHENTICATOR_ERROR"; reason: string }
+        > {
+            const errorCodeMap = {
+                SIGN_UP_NOT_ALLOWED:
+                    "Cannot sign up due to security reasons. Please try logging in, use a different login method or contact support. (ERR_CODE_007)",
+                INVALID_AUTHENTICATOR_ERROR: {
+                    // TODO: add more cases
+                },
+                WRONG_CREDENTIALS_ERROR: "The sign up credentials are incorrect. Please use a different authenticator.",
+                LINKING_TO_SESSION_USER_FAILED: {
+                    EMAIL_VERIFICATION_REQUIRED:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_013)",
+                    RECIPE_USER_ID_ALREADY_LINKED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_014)",
+                    ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_015)",
+                    SESSION_USER_ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR:
+                        "Cannot sign in / up due to security reasons. Please contact support. (ERR_CODE_016)",
+                },
+            };
+
+            const generatedOptions = await options.recipeImplementation.getGeneratedOptions({
+                webauthnGeneratedOptionsId,
+                tenantId,
+                userContext,
+            });
+            if (generatedOptions.status !== "OK") {
+                return { status: "WRONG_CREDENTIALS_ERROR" };
```

Comment:
The status is a generic one, for whatever error happens because of the Webauthn credentials flow. We could use something more specific, but it wouldn't  really make sense to the end user (they would understand and they couldn't act in any way on it).

Also, as a general note: The credential in the Webauthn recipe represents what is generated by the browser for the webauthn flow (public key). See: https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialCreationOptions.

What do you think ?

Reply:
It does make sense for the sign in/sign in options and also kind of makes sense for registerOptions, but I think INVALID_WEBAUTHN_OPTIONS_ID would make more sense here and in cases where we can know that that is the reason it failed (which is the case here I think).

Reply:
We'll be using a bit more context in the errors. Also don't need to be overly specific - something like the example above should be fine.

Reply:
done